tests/core/actionrun_test.py

```diff
@@ -335,6 +339,23 @@ def test_is_broken(self):
         self.action_run.machine.state = ActionRun.STATE_QUEUED
         assert not self.action_run.is_broken
 
+    def test_retry(self):
+        self.action_run.retries_remaining = 2
```

Comment:
There are two retries allowed, so there should be 3 failures before the action_run becomes failed, right? How does this test work? (Why are there only two EXITING updates?)

The new math in actionrun.py does look right to me, though.

Reply:
good point! retry logic actually had a subtle bug. i've moved it all to `.fail` method and fixed it.